blueocean-dashboard/src/main/js/credentials/git/GitCredentialsPicker.jsx

```diff
@@ -0,0 +1,138 @@
+    } else if (document.selection) {
+        document.selection.empty();
+    }
+}
+
+@observer
+class GitCredentialsPicker extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {};
+        this.restOrgPrefix = AppConfig.getRestRoot() + '/organizations/' + AppConfig.getOrganizationName();
+    }
+
+    componentWillMount() {
+        const { onStatus, dialog, onComplete } = this.props;
+        if (onStatus) {
+            onStatus('promptLoading');
+        }
+        Fetch.fetchJSON(this.restOrgPrefix + '/user/publickey/')
+        .then(credential => {
+            this.setState({ credential });
+            if (onStatus) {
+                onStatus('promptReady');
+            }
+            if (!dialog) {
+                onComplete(credential);
+            }
+        });
+    }
+
+    copyPublicKeyToClipboard() {
+        const textBox = this.refs.publicKey;
+        textBox.select();
+        copySelectionText();
+        clearSelection();
+        textBox.blur();
+    }
+
+    testCredentialAndCloseDialog() {
+        const { onComplete, repositoryUrl, pipeline } = this.props;
+        const fetchOptions = {
+            method: 'PUT',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({
+                repositoryUrl,
+                pipeline,
+                credentialId: this.state.credential.id,
+            }),
+        };
+        this.setState({ connectStatus: { result: 'running' } });
+        return Fetch.fetchJSON(this.restOrgPrefix + '/scm/git/validate', { fetchOptions })
+        .then(() => {
+            this.setState({
+                credentialError: null,
+                connectStatus: {
+                    result: 'success',
+                    reset: false,
+                },
+            });
+            onComplete(this.state.credential);
+        })
+        .catch(error => {
+            const message = error.responseBody ? error.responseBody.message : 'An unknown error occurred';
+            this.setState({
+                credentialError: message && t('creation.git.step1.credentials_publickey_invalid'),
+                connectStatus: {
+                    reset: true,
+                },
+            });
+        });
+    }
+
+    closeDialog() {
+        this.context.router.goBack();
+    }
+
+    render() {
+        if (!this.state.credential) {
+            return null;
+        }
+        return (
+            <div className="credentials-picker-git">
+                <p className="instructions">
+                    {t('creation.git.credentials.register_ssh_key_instructions')}{' '}
+                    <a target="jenkins-docs" href="https://jenkins.io/doc/book/blueocean/creating-pipelines/#creating-a-pipeline-for-a-git-repository">learn more</a>.
+                </p>
+                <FormElement>
+                    <textarea className="TextArea-control" ref="publicKey"
```

Comment:
This style of ref usage is deprecated:  https://facebook.github.io/react/docs/refs-and-the-dom.html#legacy-api-string-refs

Reply:
Also could you expand on what made you use the raw `textarea` element versus the JDL `TextArea` control? If there is a shortcoming on the JDL control, we should file a ticket to improve it so we don't have to dup styles in `git.less`? I'm guessing it might be the copy/paste behavior?

Reply:
There aren't complete duplicated styles in git.less but yes, it would be nice. For one it needed `readonly` but not `disabled` for clipboard copy to work right and the field to remain uneditable. It also needed a direct ref to the `textarea` component (although that could be found by traversing the DOM)